### PR TITLE
[#3378] Fix NPE in case CoAP PUT URI device id is empty

### LINKED ISSUE #3378: NPE in CoAP adapter on PUT request without device id in URI
Found this exception in CoAP adapter logs:
```
java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Unknown Source)
	at org.eclipse.hono.auth.Device.getPrincipal(Device.java:46)
	at org.eclipse.hono.auth.Device.<init>(Device.java:40)
	at org.eclipse.hono.adapter.coap.AbstractHonoResource.lambda$getPutRequestDeviceAndAuth$2(AbstractHonoResource.java:145)
	at java.base/java.util.Optional.ifPresentOrElse(Unknown Source)
	at org.eclipse.hono.adapter.coap.AbstractHonoResource.getPutRequestDeviceAndAuth(AbstractHonoResource.java:139)
	at org.eclipse.hono.adapter.coap.AbstractHonoResource.createCoapContextForPut(AbstractHonoResource.java:195)
	at org.eclipse.hono.adapter.coap.TracingSupportingHonoResource.handleRequest(TracingSupportingHonoResource.java:182)
	at org.eclipse.californium.core.server.ServerMessageDeliverer$1.run(ServerMessageDeliverer.java:129)
	at org.eclipse.hono.adapter.coap.HonoRootResource$1.lambda$execute$0(HonoRootResource.java:49)
	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
```

### PR DESCRIPTION
**Pull request by calohmn**
This fixed #3378.

### CODE REVIEW

File: adapters/coap/src/test/java/org/eclipse/hono/adapter/coap/AbstractHonoResourceTest.java [hunk 1]
```diff
@@ -117,7 +117,11 @@ void getPutRequestDeviceAndAuthSucceeds(
     @ParameterizedTest
     @CsvSource(value = {
             "true,/telemetry/OTHER_TENANT/device_1,403",
+            "true,/telemetry/OTHER_TENANT/,404",
```
Comment (calohmn) [inline adapters/coap/src/test/java/org/eclipse/hono/adapter/coap/AbstractHonoResourceTest.java:120]: Note that previously 403 was returned in this case (PUT request of an authenticated device without device ID in URI path and with non-matching tenant).
FMPOV, returning 404 fits better here.

File: adapters/coap/src/main/java/org/eclipse/hono/adapter/coap/AbstractHonoResource.java [hunk 2]
```diff
@@ -138,11 +138,14 @@ protected Future<RequestDeviceAndAuth> getPutRequestDeviceAndAuth(final CoapExch
         Optional.ofNullable(TracingSupportingHonoResource.getAuthenticatedDevice(exchange))
             .ifPresentOrElse(
                     authenticatedDevice -> {
-                        final String tenantFromUri = Optional.ofNullable(requestedResource.getTenantId())
+                        final String tenantId = Optional.ofNullable(requestedResource.getTenantId())
```
Comment (sophokles73) [inline adapters/coap/src/main/java/org/eclipse/hono/adapter/coap/AbstractHonoResource.java:141]: IMHO we should also return a failed future (404) if the request URI does not contain a tenant ID as the CoAP user guide seems to require it for a PUT request, shouldn't we?

Comment (calohmn) [reply adapters/coap/src/main/java/org/eclipse/hono/adapter/coap/AbstractHonoResource.java:141]: A tenant ID is required for an _unauthenticated_ device request. This case is handled in the block starting on line 157.
There is already a 404 returned in this case (see also test method param `"false,/event//device_1,404",`).

Comment (sophokles73) [reply adapters/coap/src/main/java/org/eclipse/hono/adapter/coap/AbstractHonoResource.java:141]: Oh yes, my bad, the PUT operation is also used by an authenticated gateway ...

**Review by sophokles73 — APPROVED** [review]
LGTM

**Comment by sophokles73** [comment]
@calohmn does this also affect 2.0.1?

**Comment by calohmn** [comment]
yes, I want to merge the fix also in 2.0.x.

**Pull request merged**